traits/observers/_metadata_filter.py

```diff
@@ -0,0 +1,61 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" A filter to be used with FilteredTraitObserver for observing traits
+with a given metadata and value filter. The filter is defined here so that two
+observers created using the same parameters compare equally.
+"""
+
+
+class MetadataFilter:
+    """ Callable to be used with FilteredTraitObserver for filtering traits
+    using defined metadata.
+
+    If the metadata is not defined, the filter will return False.
+
+    Attributes
+    ----------
+    metadata_name : str
+        Name of the metadata to filter traits with.
+        If the trait does not have the requested metadata, this filter will
+        return false.
+    value : callable(value) -> boolean
+        The callable must accept a single argument which is
+        the metadata value on the trait and return true if the trait
+        is to be observed. For removing an existing observer, the ``value``
+        callables must compare equally. The callable must be hashable as well.
+    """
+
+    def __init__(self, metadata_name, value):
+        self.metadata_name = metadata_name
+        self.value = value
+
+    def __call__(self, name, trait):
+        # hasattr on an CTrait always return true, and the
+        # value would be None if it is not defined.
+        if self.metadata_name in trait.__dict__:
+            value = getattr(trait, self.metadata_name)
+            return self.value(value)
+        return False
```

Comment:
This `value` filter is more than necessary for supporting the `"+metadata_name"` feature in the mini-language.
On the one hand, it is easy to implement, and is consistent with `HasTraits.traits` support on filtering metadata using the values. On the other hand, this is a feature I am not sure if we have a use case for (but maybe we do, they don't exist because users have been working around its nonexistence.)

Furthermore, the filtering here is not the exactly same as that in `HasTraits.traits`: `HasTraits.traits` does not differentiate when a metadata is defined or not. If the metadata is undefined, it assumes it existed with `None` as the value and passes it to the filter and I don't think we wanted the same behaviour here.

Reply:
We could remove this `value` and deem this `MetadataFilter` for checking for metadata existence only, for now. If one really needs to filter the metadata value, `filter_(lambda name, trait: trait.my_metadata_name is not None and trait.my_metadata_name > 10.0)` can still be defined for that use case (note that `is not None` check), and if it becomes common enough, add this value filter back is easy.

Reply:
Agreed; that sounds good to me.

Reply:
`value` is removed. This is now ready for review again.